src/main/scala/dataclass/Macros.scala

```diff
@@ -185,14 +188,30 @@ private[dataclass] class Macros(val c: Context) extends ImplTransformers {
                 .map { param =>
                   q"code = 37 * code + this.${param.name}.##"
                 }
-              Seq(
-                q"""override def hashCode: _root_.scala.Int = {
+              if (generatedCachedHashCode)
+                Seq(
+                  q"""private var __hasHashCode: Boolean = false""",
```

Comment:
I'm not sure this is thread-safe… `__hasHashCode` and `__hashCode` should be marked as `volatile` at least, so that concurrent accesses don't go wrong.

I'm wondering whether we could even drop `__hasHashCode` altogether, and use a sentinel value for `__hashCode` to know whether it's been set or not. Like assume `0` means it wasn't set, and if the computed hash code is 0, write 1 instead (which shouldn't be a problem with the semantics of `hashCode`).

Reply:
Yes, we can do the sentinel value. I thought about it, and I was pretty neutral which one to choose. That way we'll save some space.

I'm very hesitant to add volatile here, it might have very bad performance implications.

If I understand things correctly you can't observe half a written int on the jvm, writes and reads are atomic. As such the worst case here is that we recalculate hashcode very rarely. I'd say that is much better than slow reads always. Agree?

I'll update the PR later today

Reply:
Let's try without `volatile` then (I'm a bit nervous about this though 😅)

Reply:
maybe this will calm you? https://stackoverflow.com/questions/41704185/is-javas-string-hashcode-function-thread-safe-if-its-cache-setter-does-not-us

Reply:
> maybe this will calm you? https://stackoverflow.com/questions/41704185/is-javas-string-hashcode-function-thread-safe-if-its-cache-setter-does-not-us

I wouldn't have stated that this way :/ [This page](https://docs.oracle.com/javase/tutorial/essential/concurrency/atomic.html) linked there explicitly states that reference variables and most primitive types are read and written atomically, so this works.